토큰 재발급 기능 구현

GangHub1970 opened the pull request:
## ⭐ Key Changes

- 권한이 필요한 API 요청을 보내는 인스턴스(`authInstance`)와 권한이 필요하지 않은 API 요청을 보내는 인스턴스(`instance`)로 구분했습니다.
  - `accessToken` 만료 시 재발급 요청을 보내는 로직을 에러 코드(`401`)로 구분해서 진행하고 있습니다. 그런데 권한이 필요한 API의 경우 `401`에러는 모두 `accessToken`이 만료되었다는 에러인데, 권한이 필요없는 API는 다른 경우(ex. 유효성 검사)에도 `401`에러가 발생해서 인스턴스를 구분하는 방법을 사용했습니다.

- 사용자의 최초 접근시 토큰을 재발급 받습니다.
  - 사용자의 `accessToken` 유무에 따라 재발급 요청을 보낼지 말지 판단합니다.
  - 새로고침, URL 직접 수정을 통한 페이지 이동에는 `accessToken`이 없는 경우 재발급 요청을 보냅니다.
  - 만약 재발급을 받는다면 그 시간동안은 하위 컴포넌트에서 다른 API 요청이 가지 않게 로딩화면을 보여줍니다.

- `accessToken`을 저장할 때 만료 시간을 추가했습니다.
  - 만료 시간을 반환하는 util 함수(`getTokenExpDate`)를 구현했습니다.
  - 혹시 `accessToken`에 있는 정보를 사용할 곳이 있을 수 있을것 같아 토큰의 페이로드를 반환하는 함수(`getTokenPayload`)도 구현했습니다. 추후 사용되지 않으면 삭제해도 될 것 같습니다.

### 🚀 추가 변경사항

- `accessToken`을 쿠키에 저장하지 않고 전역 상태(`accessTokenAtom`)에 저장하는 방법으로 변경했습니다.
  - `isLoggedIn`은 `accessToken`의 상태에 따라 변경되는 값이기 때문에 `selector`로 구현했습니다.
  - `accessToken`을 전역 상태로 저장하면서 컴포넌트가 아닌 부분(`authInstance.ts`)에서 토큰값을 사용하는데 어려움이 있어 `AuthInterceptor` 컴포넌트를 만들고 `interceptor`를 설정했습니다.

## 🖐️To reviewers

- 저번 회의 때 `accessToken` 만료 직전의 API 요청에서 재발급 요청을 보낸다고 말씀드렸는데, `401`에러로 만료를 확인하고 재발급받는 방법으로 변경했습니다.
  - 이유는 전자의 경우 매 요청에서 토큰의 만료시간을 확인하고 refresh 요청을 보내야, 후자의 경우 에러가 나는 경우에만 요청을 보내면 되기 때문입니다.

- 인스턴스 분리로 인해 기존 코드에서 변경해야 하는 부분이 생길 것 같습니다. 하지만 이 방법이 별로라는 피드백을 받으면 다른 방법을 찾아보겠습니다.

- API 요청 함수에 반환값은 지정하지 않았었다는 피드백을 적용시켜 `login` 함수에서도 삭제했습니다.

## 📌issue

close #23

GangHub1970 committed:
feat: 권한이 필요한 api에서 401에러 발생 시 리프레시

- 권한이 필요한 api와 필요 없는 api의 401 에러를 구분하기 위해 인스턴스 구분
- 리프레시 요청에서 400, 401 에러 발생 시 로그인 페이지로 이동

GangHub1970 committed:
feat: 토큰 만료 날짜 반환 util 함수 생성

GangHub1970 committed:
feat: 최초 서비스 접근 시 refresh 요청

- 최초 서비스 접근 후 새로고침 혹은 페이지 이동이 한 번 있고난 후부터 정상적으로 로딩을 거치지 않음

GangHub1970 committed:
feat: 서비스 접근시 토큰 재발급

- 서비스에 접근하면 refresh 요청을 통해 토큰을 재발급 받습니다.
- 새로고침, URL 수정을 통한 접근시에는 매번 refresh 요청을 보냅니다.

GangHub1970 committed:
refactor: API 요청 함수에 반환 타입을 삭제

- 코드 통일성을 유지하기 위해 삭제했습니다.

itsmesunky reviewed (changes requested):
전반적으로 고민한 흔적이 많이 느껴지는 코드였습니다! 고생하셨습니다. 👍👍

itsmesunky commented on src/apis/@core/authInstance.ts:
```diff
@@ -0,0 +1,64 @@
+          const { status } = err;
+
+          // 리프레시 요청에서도 에러가 발생하는 경우는 로그인 페이지로 이동
+          if (status === 400 || status === 401) {
```
로그인 창으로 바로 이동하기보다는 `alert`을 활용해 알림 메시지를 먼저 표시하면 사용자 입장에서 상황을 이해하기 더 쉬워지고, UX가 개선될 거 같습니다!

```typescript
alert("세션이 만료되었습니다. 다시 로그인해 주세요.");
```

GangHub1970 replied:
피드백 감사드립니다! 피드백 반영해서 사용자에게 `alert`로 알려주고 페이지 이동되게 수정했습니다.
현재는 `alert`로 구현했고, 디자인 들어가면서 수정할 때는 토스트나 모달로 구현하면 될 것 같습니다 👍

itsmesunky commented on src/apis/@core/authInstance.ts:
권한이 필요 없는 API 호출 시에도 `401` 에러가 발생할 가능성이 있어서 프론트에서 구분하기가 애매하긴 하네요...
혹시 이 부분에 대해 백엔드측에서  `401` 에러를 다른 상태 코드로 구분해서 처리할 수 있는 방법은 없다고 하셨을까요?

만약, 그게 불가능하거나 공수가 많이 드는 작업이라면 현재처럼 인스턴스를 분리하는 방식으로 진행하는 게 가장 나을 것 같긴 합니다!

iamkjw77 replied:
권한이 필요없는 API에도 401에러가 발생하는게 이상하네요...🤔
 이번주 전체회의에서 권한이 필요한/필요하지 않은 API를 구분 요청하고, 
 권한이 필요하지 않은 API에서는 401에러가 나지않도록 요청하는게 좋을 것 같아요.
 
그럼 2개의 케이스(권한이 필요한/필요하지 않은)로 나뉘어서, 위 방법처럼 인스턴스를 구분하는 것이 좋을 것 같습니다!

GangHub1970 replied:
백엔드에 권한이 필요없는 API에 `401`에러를 다른 에러코드로 변경해줄 수 있냐는 요청은 해보지 않았습니다ㅎㅎ

제 생각에는 이미 API가 저렇게 설계되어 있고, 인스턴스를 나누면 권한이 필요없는 API 요청할 때는 프론트에서도 요청 `header`에 토큰 값을 안넣어도 되니까 인스턴스를 구분하는게 더 괜찮겠다고 생각했습니다!!

권한이 필요없는 API에서 `401`에러가 나는 경우는 로그인할 때 `비밀번호 인증에 실패한 경우`, `보안 코드 인증에 실패한 경우` 이런 경우가 있고 더 있는 것 같습니다.

iamkjw77 commented on src/apis/@core/authInstance.ts:
```diff
@@ -0,0 +1,64 @@
+    } = err;
+
+    // 권한이 필요한 api 요청에서 access token 에러가 발생하는 경우
+    if (status === 401) {
```
401에러가 계속해서 발생하는 경우에는 무한루프에 빠질 수 있을 것 같아요.
재시도하는 count를 정해서 그 이상 넘어가는 경우에는 
로그아웃 후 로그인 페이지로 이동하는 예외처리가 필요할 것 같습니다😀

GangHub1970 replied:
`refresh` 요청에서 발생하는 에러는 처리하기 위해 `try...catch`문으로 처리했는데, 혹시 `401`에러가 계속해서 발생하는 경우가 어떤 경우가 있을까요??

iamkjw77 replied:
무한루프는 AT가 쿠키에 저장되어 있어서 refresh 요청에서 에러가 발생한 경우입니다. 이 경우 AT 저장방법을 달리하면 해결될 듯 합니다.
제 생각에는 아래와 같이 작동할 것 같은데, 잘못된 부분이 있으면 첨언 부탁드립니다.

1. refresh 요청 실패로 로그인 페이지로 리다이렉트(아직 쿠키에 AT가 저장됨)
2. isLoggedIn이 true가 되어, 컴포넌트 측에서 권한이 필요한 API 요청이 있다면 401에러 반환
3. 로그인 페이지로 리다이렉트
... (1번 상황으로 반복)

위 같은 401에러가 무한루프로 반복되는 상황이 있을 수 있을 것 같습니다.
이 문제는 AT을 오롯이 recoil 상태에만 저장하면 리다이렉트 될 때, AT가 날라가면서 해결될 것 같습니다.

GangHub1970 replied:
AT가 있다면 RT도 있을거고, 그러면 refresh 요청이 에러가 발생하지 않을 것으로 생각됩니다.
그리고 만약 AT가 있는 상태로 로그인 페이지로 왔다고 해도 로그인 페이지로 이동한 후에는 권한이 필요한 API 요청이 없는게 맞다고 생각이 되네요..!
무한루프 상황은 아직 이해가 잘 안되긴한데.. 제 설명에도 잘못된 부분이 있다면 첨언 부탁드립니다.

그래도 보안상 AT를 전역 상태로 관리하는게 맞다고 생각해 수정했습니다!

iamkjw77 replied:
refresh 요청이 실패한 상황(서버 에러)을 가정해서 생각했습니다.
AT가 있는 상태로 로그인 페이지 또는 로그인 페이지에서 다른 페이지로 이동 시에 
쿠키에 AT가 존재하기 때문에 권한이 있는 API 요청이 있을 수 있다고 생각했습니다.

refresh 요청이 성공하거나 AT를 전역상태로 빼면 제가 생각한 상황도 의미가 없을 것 같습니다!

iamkjw77 commented on src/hooks/auth/index.ts:
```diff
@@ -20,8 +21,11 @@ export const useLogin = ({ loginErrorHandler }: UseLoginParams) => {
+      const expirationDate = getTokenExpDate(accessToken);
 
-      setCookie("accessToken", accessToken);
+      setCookie("accessToken", accessToken, {
```
accessToken을 쿠키에 저장할 경우, 페이지를 리프레시 하거나 창을 닫고 다시 접속할 때도 해당 토큰이 유지되어서 XSS와 CSRF 공격에 취약한 걸로 알고 있습니다. 

AT를 쿠키에서 가져와 저장하는 방법대신, 리코일 전역상태에 바로 저장하는 방법은 어떨까요?

GangHub1970 replied:
오...좋은 것 같습니다!
말씀해주신 방법대로 변경하면 로그인 되어있는 상황에도 `refresh` 요청을 보낸다는 단점이 생기겠지만, 새로고침했을 때 한번만 발생하는 문제이니 보안을 챙기는 방법이 더 좋을 것 같습니다. 👍

iamkjw77 reviewed (changes requested):
고생하셨습니다👍

GangHub1970 committed:
feat: interceptor의 refresh 요청에서 에러가 뜨는 경우 alert 처리

GangHub1970 committed:
refactor: access token을 recoil로 관리

- AT를 쿠키에 저장했을 때 발생하는 보안 취약점을 보완하기 위해 recoil로 관리
- AppInitializer => TokenRefresher 로 네이밍 변경
- 기존 isLoggedIn 상태를 auth로 변경하여 사용자의 인증 상태를 저장

GangHub1970 committed:
refactor: 사용자 인증 상태 분리

- isLoggedIn이 accessToken의 유무에 따라 정해지기 때문에 accessToken을 atom으로 만들고 isLoggedIn을 selector로 구현

iamkjw77 commented on src/recoil/accessToken/index.ts:
폴더구조 컨벤션에 맞춰서 `/src/recoil/auth/accessToken/index.ts`로 맞추는 것이 좋을 것 같습니다.
(아래 inLoggedIn도 동일합니다!)

GangHub1970 committed:
Merge branch 'develop' into feature/refresh

# Conflicts:
#	src/apis/auth/index.ts
#	src/types/auth/index.ts

GangHub1970 committed:
style: accessToken 전역 상태 폴더 이동

iamkjw77 commented on src/hooks/auth/index.ts:
`/src/hooks/auth/useLogin.ts`, `/src/hooks/auth/useRefresh.ts` 같이 auth에서 사용하는 훅을 두 개의 파일로 나눈 후에 파일 네이밍을 use~로 시작하면 어떤 커스텀 훅이 있는지 바로 볼 수 있어서 좋을 것 같습니다.

GangHub1970 replied:
훅이 많이지면 찾기 어려울것 같아서 한 곳에서 관리했는데 폴더가 관심사별로 나눠지니까 파일을 분리하는것도 좋을 것 같습니다!

수정했습니다.👍

GangHub1970 committed:
style: 커스텀 훅 파일 분리

- 한 파일에서 하나의 훅만 관리하도록 분리해서 파일명을 보고 무슨 훅인지 알 수 있게 수정
